docs/Macros.md

```diff
@@ -98,4 +98,103 @@ the resulting code. `macro-log` is useful for tracing your macro, a form of
 
 ## Inner Workings
 
-TODO
+The Carp compiler is split in a few different stages. The diagram below
+illustrates the flow of the compiler.
+
+![The compiler passes](./compiler-passes.svg)
+
+The dynamic evaluator is arguably one of the most central pieces of the Carp
+compiler. It orchestrates macro expansion, borrow checking, and type inference,
+as it encounters forms that have requirements for these services, such as
+function definitions, variables, or `let` bindings.
+
+As such, the evaluator serves many purposes. While you might start out trying
+to gain an understanding of macro expansion, understanding the evaluator will
+give you a lot of insight into how Carp works generally.
+
+The most tried-and-true starting point for understanding the dynamic evaluator
+is `eval` in [`src/Eval.hs`](/src/Eval.hs).
+
+### Data Structures
+
+The type signature of `eval` is as follows:
+
+```haskell
+eval :: Context -> XObj -> IO (Context, Either EvalError XObj)
+```
+
+Thus, to understand it, we’ll have to understand at least `Context`, `XObj`,
+and `EvalError`. The types `IO` and `Either` are part of the Haskell standard
+library and will not be covered extensively—please refer to your favorite
+tool for Haskell documentation (we recommend [Stackage](https://stackage.org))
+to find out more about them.
+
+All data structures that are discussed here are defined in
+[`src/Obj.hs`](/src/Eval.hs).
+
+#### `XObj`
+
+`XObj` is short for “`Obj` with eXtras”. `Obj` is the type for AST nodes in
```

Comment:
Amazing. This sort of information is just what I need, for example, to add proper immutability checking :)